File: src/memory.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub fn growCapacity(cap: usize) usize {
    if (cap < 8) {
        return 8;
    }
    return 8;
}

pub fn growArray(comptime T: type, allocator: Allocator, ptr: ?[]T, old_size: usize, new_size: usize) !?[]T {
    return try reallocate(T, allocator, ptr, old_size, new_size);
}

pub fn freeArray(comptime T: type, allocator: Allocator, ptr: ?[]T, old_size: usize) !?[]T {
    return try reallocate(T, allocator, ptr, old_size, 0);
}

pub fn reallocate(comptime T: type, allocator: Allocator, ptr: ?[]T, old_size: usize, new_size: usize) !?[]T {
    _ = old_size;
    if (new_size == 0 and ptr != null) {
        allocator.free(ptr.?);
        return null;
    }
    if (new_size == 0 and ptr == null) {
        return null;
    }
    return allocator.alloc(T, new_size) catch |err| {
        std.debug.panic("unable to allocate!!!. {any}", .{err});
    };
}

File: c/src/memory.c
#include <stdio.h>
#include <stdlib.h>

#include "memory.h"

void* reallocate(void* pointer, size_t old_size, size_t new_size, char* file, int line) {
    (void)old_size;
    if (new_size == 0) {
        free(pointer);
        return NULL;
    }
    void* result = realloc(pointer, new_size);
    if (result == NULL) {
        fprintf(stderr, "Unable to realloc %s:%d\n", file, line);
        exit(1);
    }
    return result;
}
